relay-server/src/endpoints/common.rs

```diff
@@ -433,11 +439,63 @@ where
                         Err(BadStoreRequest::PayloadError(StorePayloadError::Overflow))
                     }
                 }))
-                .and_then(move |(envelope, rate_limits)| {
+                .and_then(clone!(
+                    event_manager,
+                    project_manager,
+                    event_id,
+                    |(envelope, rate_limits)| {
+                        // do dynamic sampling on transactions
+                        let trace_context = match envelope.trace_context() {
+                            None => {
+                                // if we don't have a trace context we can't do dynamic sampling so stop.
+                                return Box::new(Ok((envelope, rate_limits, None)).into_future())
+                                    as ResponseFuture<
+                                        (Envelope, RateLimits, Option<Addr<Project>>),
+                                        _,
+                                    >;
+                            }
+                            Some(trace_context) => trace_context,
+                        };
+                        // Sample and potentially remove transactions from the envelope.
+                        // We only do this if the envelope contains only transactions,
+                        // The reason for that is that in case of envelopes containing only
+                        // transactions we have a chance to end processing here (if the transactions
+                        // are sampled out).
+                        // If the envelope contains other items (beyond transactions then we cannot
+                        // shortcut the processing here so we'll do it after we queue the envelope).
+                        let response = project_manager
+                            .send(GetProject {
+                                public_key: trace_context.public_key,
+                            })
+                            // deal with mailbox errors
+                            .map_err(BadStoreRequest::ScheduleFailed)
+                            // do the fast path transaction sampling (if we can't do it here
+                            // we'll try again after the envelope is queued)
+                            .and_then(|project| {
+                                let event_id = envelope.event_id();
+                                sample_transaction(envelope, Some(project.clone()), true).then(
```

Comment:
_nit_: This `.then` can be moved to the outer closure.

Reply:
refactored, I think it looked better with one top level and_then for dynamic sampling rather then 2.